### PR TITLE
Upgrade typescript version

### PR DESCRIPTION
**Pull request by fa93hws**
Latest version typescript has more strict type check.

### CODE REVIEW

File: src/servicesHost.ts [hunk 1]
```diff
@@ -670,9 +670,19 @@ function addCache(servicesHost: typescript.ModuleResolutionHost) {
       const cache = createCache<ReturnType<typeof originalFunction>>(
         originalFunction
       );
-      servicesHost[
-        functionToCache
-      ] = cache.getCached as typescript.ModuleResolutionHost[CacheableFunction];
+      switch (functionToCache) {
```
Comment (johnnyreilly) [inline src/servicesHost.ts]: It's cool that you can make the compiler play ball by doing this.  However, I kind of don't like it as we're essentially writing more code to get round a problem that the compiler is presenting us with....

@andrewbranch can you think of a nicer approach than the one suggested here? (Funnily enough it looks like this compiled fine prior to TS 3.5... Something changed there I'm guessing)

Comment (johnnyreilly) [reply src/servicesHost.ts]: Although maybe the approach suggested here is better as it's certainly clearer... It's just the boilerplate I object to I guess.  It's one of those hard to understand compiler messages:

```
Type '((fileName: string) => boolean) | ((path: string) => string) | ((directoryName: string) => boolean) | undefined' is not assignable to type '((fileName: string) => boolean) & ((directoryName: string) => boolean) & ((path: string) => string)'.
  Type 'undefined' is not assignable to type '((fileName: string) => boolean) & ((directoryName: string) => boolean) & ((path: string) => string)'.
    Type 'undefined' is not assignable to type '(fileName: string) => boolean'.ts(2322)
```

**Comment by andrewbranch** [comment]
Yeah, I don't think the loop is really worth it. Personally I’d probably go with something like

```ts
function addCache(servicesHost: typescript.ModuleResolutionHost): {
  moduleResolutionHost: typescript.ModuleResolutionHost,
  clearCache: () => void,
} {
  const clearCacheFunctions: Action[] = [];
  return {
    moduleResolutionHost: {
      ...servicesHost,
      fileExists: createCache(servicesHost.fileExists),
      directoryExists: servicesHost.directoryExists && createCache(servicesHost.directoryExists),
      realpath: servicesHost.realpath && createCache(servicesHost.realpath),
    },
    clearCache: () => clearCacheFunctions.forEach(clear => clear()),
  };

  function createCache<TOut>(originalFunction: (arg: string) => TOut) {
    const cache = new Map<string, TOut>();
    clearCacheFunctions.push(() => cache.clear());
    return function getCached(arg: string) {
      let res = cache.get(arg);
      if (res !== undefined) {
        return res;
      }

      res = originalFunction(arg);
      cache.set(arg, res);
      return res;
    };
  }
}
```

and then at the usage site

```ts
  let clearCache: null | (() => void) = null;
  let moduleResolutionHost: ModuleResolutionHost = {
    fileExists,
    readFile: readFileWithFallback,
    realpath: compiler.sys.realpath,
    directoryExists: compiler.sys.directoryExists,
    getCurrentDirectory: compiler.sys.getCurrentDirectory,
    getDirectories: compiler.sys.getDirectories
  };

  if (enableFileCaching) {
    const cached = addCache(moduleResolutionHost);
    clearCache = cached.clearCache;
    moduleResolutionHost = cached.moduleResolutionHost;
  }
```

I guess that’s slightly more wordy than what’s proposed here, but doesn’t contain any unsafe type assertions.

**Comment by johnnyreilly** [comment]
Thanks for the suggestion @andrewbranch!

@fa93hws do you fancy tweaking your PR with the suggested implementation?

**Comment by andrewbranch** [comment]
I feel like I should add that there are of course lots of different solutions that largely come down to personal preference of coding style. Avoiding unnecessary type assertions is a worthwhile goal, but I’m sure mine is not the most elegant possible solution.

**Comment by fa93hws** [comment]
No problem!

File: package.json [hunk 2]
```diff
@@ -88,7 +88,7 @@
     "rimraf": "^2.6.2",
     "tslint": "^5.11.0",
     "tslint-config-prettier": "^1.15.0",
-    "typescript": "^3.1.1",
+    "typescript": "^3.5.1",
     "webpack": "^4.5.0",
```
Comment (johnnyreilly) [inline package.json]: Shall we go for TypeScript 3.5.2 since it's out?

Comment (fa93hws) [reply package.json]: No problem!

**Comment by johnnyreilly** [comment]
Thanks!

**Pull request merged**